packages/api/src/codegen/typescript.ts

```diff
@@ -0,0 +1,357 @@
+    }));
+  }
+
+  /**
+   * Convert a JSON Schema object into a readily available TypeScript type or interface along with
+   * any `$ref` pointers that are in use and turn those into TS types too.
+   *
+   * Under the hood this uses https://npm.im/json-schema-to-typescript for all composition and
+   * conversion.
+   *
+   * @param schema
+   * @param name
+   */
+  async convertJSONSchemaToTypescript(schema: JSONSchema, name: string) {
+    // Though our JSON Schema type exposes JSONSchema4, which `json-schema-to-typescript`, it won't
+    // accept our custom type.
+    const ts = await compile(schema as any, name, {
+      bannerComment: '',
+    });
+
+    let primaryType: string;
+    const tmp = this.project.createSourceFile(`${name}.types.tmp.ts`, ts);
+    const declarations = tmp.getExportedDeclarations();
+
+    Array.from(declarations.keys()).forEach(declarationName => {
+      if (!primaryType) {
+        primaryType = declarationName;
+      }
+
+      declarations.get(declarationName).forEach(declaration => {
+        this.types.set(declarationName, declaration.getText());
+      });
+    });
+
+    this.project.removeSourceFile(tmp);
+
+    return {
+      primaryType,
+    };
+  }
+
+  /**
+   * Scour through the current OpenAPI definition and compile a store of every operation, along
+   * with every HTTP method that's in use, and their available TypeScript types that we can use,
+   * along with every HTTP method that's in use.
+   *
+   */
+  async loadOperationsAndMethods() {
+    const operations: Record</* operationId */ string, OperationTypeHousing> = {};
+    const methods: string[] = [];
+
+    await Promise.all(
+      Object.entries(this.spec.getPaths()).map(async ([, ops]) => {
+        await Promise.all(
+          Object.entries(ops).map(async ([method, operation]: [HttpMethods, Operation]) => {
+            if (!(method in methods)) {
+              methods.push(method);
+            }
+
+            const operationId = operation.getOperationId();
+            const params = await this.getParameterTypesForOperation(operation, operationId);
+            const responses = await this.getResponseTypesForOperation(operation, operationId);
+
+            if (operation.hasOperationId()) {
+              operations[operation.getOperationId()] = {
+                types: { params, responses },
+                operation,
+              };
+            }
+          })
+        );
+      })
+    );
+
+    return {
+      operations,
+      methods,
+    };
+  }
+
+  /**
+   * Compile the parameter (path, query, cookie, and header) schemas for an API operation into
+   * usable TypeScript types.
+   *
+   * @param operation
+   * @param operationId
+   */
+  async getParameterTypesForOperation(operation: Operation, operationId: string) {
+    let params;
+    const paramSchemas = operation.getParametersAsJsonSchema({
+      mergeIntoBodyAndMetadata: true,
+      retainDeprecatedProperties: true,
```

Comment:
These two options were added in https://github.com/readmeio/oas/pull/617 to allow me to merge all metadata-categorized parameters (query, header, cookie, path) into a single `metadata` JSON Schema entry -- and ultimately a single TS type for that parameter.